keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+  of the application
+- Jack Roper, [How to Delete Pods from a Kubernetes Node with Examples](https://spacelift.io/blog/kubectl-delete-pod), 05 Jul, 2023 - also
+  discusses the problem of blocking PDBs and offers several workarounds. Similar to others also
+  offers a force deletion, but also a less destructive method of scaling up the application.
+  However, this also interferes with application deployment and has to be supported by the
+  application.
+
+### Cluster Autoscaler
+
+Accepts a `drain-priority-config` option, which is similar to Graceful Node Shutdown in that it
+gives each priority a shutdown grace period. Also has a `max-graceful-termination-sec` option for
+pod termination and a `max-pod-eviction-time` option after which the eviction is forfeited.
+
+Each pod is first analyzed to see if it is drainable. Part of the logic is similar to kubectl and
+its drain filters (see [Cluster Autoscaler rules](https://github.com/kubernetes/autoscaler/blob/554366f979b11aeb82df335a793e4d7a1acfadb4/cluster-autoscaler/simulator/drainability/rules/rules.go#L50-L77)):
+- Mirror pods are skipped.
+- Terminating pods are skipped.
+- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
+  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
+- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
+  (the check can be modified with the `skip-nodes-with-system-pods` option).
+- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
+- Pods with local storage are blocking unless they have a
+  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
+  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
+  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
+- Pods with PDBs that do not have `disruptionsAllowed` are blocking.
+
+This can be enhanced with other rules and overrides.
+
+It uses this logic to first check if all pods can be removed from a node. If not, it will report
+those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
+to highest priority. This may include DaemonSet pods.
+
+### kubelet
+
+Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is
+not possible to rely solely on this feature as a go-to solution for graceful node and workload
+termination.
+
+- The Graceful Node Shutdown feature is not application aware and may prematurely disrupt workloads 
+  and lead to data loss.
+- The kubelet controls the shutdown process using Dbus and systemd, and can delay (but not entirely
+  block) it using the systemd inhibitor. However, if Dbus or the kubelet is restarted during the
```

Comment:
That sounds like a bug in the kubelet, no?

Reply:
AFAIK this is an inherent problem of the current design and not something that can be solved entirely by kubelet. To solve this they need to store the intent, which would be done through NodeMaintenance in this proposal.

Reply:
if kubelet is restarted how the intent will be stored? how it will reconcile after start?

Reply:
It would first have to reconnect to the apiserver, to obtain the NodeMaintenance or [Node](https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631444470) object.